### ISSUE TITLE
sort() does not preserve tensor dtype

### ISSUE THREAD

**Issue by pennyandtrie** [tinygrad/tinygrad#17113]
```python
from tinygrad import Tensor
from tinygrad.dtype import dtypes

bef_sort = Tensor([3,1,2], dtype=dtypes.uint8)
aft_sort = bef_sort.sort()[0]

print(f'{bef_sort.dtype=}\n{aft_sort.dtype=}')
``` 

produces:

bef_sort.dtype = dtypes.uchar
aft_sort.dtype  = dtypes.int

Is that intended? I would expect sorted tensor dtype to be the same as before sorting.
If it is a bug, let me know - I have a patch

**Comment by dyKiU** [comment]
same for other ops, as well - not sure if matters in the grand scheme of things?

```
>>> from tinygrad import Tensor
>>> from tinygrad.dtype import dtypes
>>> x = Tensor([1, 1, 0, 1], dtype=dtypes.float16)
>>> for name, result in {
...       "cumsum": x.cumsum(0),
...       "cumprod": x.cumprod(0),
...       "cummax": x.cummax(0)[0],
...       "cummin": x.cummin(0)[0],
...       "logcumsumexp": x.logcumsumexp(0),
...   }.items():
...       print(name, result.dtype)
... 
cumsum dtypes.half
cumprod dtypes.float
cummax dtypes.float
cummin dtypes.float
logcumsumexp dtypes.float
```

torch baseline
```
>>> import torch
>>> tx = torch.tensor([1, 1, 0, 1], dtype=torch.float16)
>>> print(tx.cumprod(0).dtype)
torch.float16
>>> print(tx.cummax(0).values.dtype)
torch.float16
>>> print(tx.cummin(0).values.dtype)
torch.float16
>>> print(torch.logcumsumexp(tx, 0).dtype)
torch.float16
```

**Comment by pennyandtrie** [comment]
imho the culprit is here:

https://github.com/tinygrad/tinygrad/blob/46b82d47555965fe2585753d6c5a98a9218c0dfd/tinygrad/mixin/op.py#L280-L289

more specifically L288:

```python
    if value is not Invalid: base = base.cast(least_upper_dtype(base.dtype, dtypes.from_py(value)))
``` 


When **_pad_constant()** is used as 'true' padding, e.g. in **pad(),** than casting is reasonable - **tensor dtype** might be of smaller bitsize than **value dtype** (upcasting needed).

However when **_pad_constant()** is used only as a 'helper', e.g. in **sort()**, than there is no point in casting

Possible patch is to add something like **preserve_dt** in **_pad_constant()**:

```python
def _pad_constant(self, pX, value:ConstType, preserve_dt=False) -> Self:
``` 
and change L288 from

```python
if value is not Invalid: base = base.cast(least_upper_dtype(base.dtype, dtypes.from_py(value)))
``` 
to 
```python
if value is not Invalid: base = base if preserve_dt else base.cast(least_upper_dtype(base.dtype, dtypes.from_py(value)))
```

and then whenever **_pad_const()** is used only as a 'helper' like in **sort()**:

 https://github.com/tinygrad/tinygrad/blob/46b82d47555965fe2585753d6c5a98a9218c0dfd/tinygrad/mixin/op.py#L919

set preserve_dt=True:

```python
x = x._pad_constant(pads, x.dtype.min if descending else x.dtype.max, preserve_dt=True).unflatten(dim, (2,)*n_stages)
```

**Comment by pennyandtrie** [comment]
@chenyuxyz sorry for being PITA but are you sure #17125 is correct fix?

Now **pad() behaviour changed**, e.g. when padding value is out of bound of tensor's dtype.

**Before fix**: original tensor dtype was upcasted; padding reflected provided value
**After fix**: original tensor dtype remains the same; padding does not reflect provided value (overflows)

e.g.:

```python
t0=Tensor([0], dtype=dtypes.uint8)
t1=t0.pad((1,1),value=256)
print(t1.numpy())
```
after fix: array([0, 1, 0], dtype=uint8)
before fix: array([257,   1, 257], dtype=int32)

tbh pytorch throws error when value is out of bounds of tensor's underlying dtype, so just want to know what is expected behaviour for tinygrad

best regards and once again sorry for the problem
